File: CreatorAI/Services/SupabaseService.swift
import Foundation
import Supabase

/// Supabase client for database and storage.
/// Auth is handled by Firebase — Supabase stores data only (same as Android).
final class SupabaseService {
    static let shared = SupabaseService()

    private static let supabaseURL = "https://xxcllmzflnwzqiaslyou.supabase.co"
    private static let supabaseAnonKey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Inh4Y2xsbXpmbG53enFpYXNseW91Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzQyMDM1NDcsImV4cCI6MjA4OTc3OTU0N30.VbAv5UEC6n1ZyXFvjLshVOp9ZKN5_Rq8fEznoEqFsKY"

    let client: SupabaseClient

    private init() {
        client = SupabaseClient(
            supabaseURL: URL(string: Self.supabaseURL)!,
            supabaseKey: Self.supabaseAnonKey
        )
    }

    // MARK: - App Users (Firebase UID + FCM token)

    private struct AppUserRow: Encodable {
        let id: String
        let email: String?
        let display_name: String?
        let avatar_url: String?
        let platform: String
    }

    func upsertUser(userId: String, email: String?, displayName: String? = nil, avatarUrl: String? = nil) async {
        let row = AppUserRow(
            id: userId,
            email: email,
            display_name: displayName,
            avatar_url: avatarUrl,
            platform: "ios"
        )
        do {
            try await client.from("app_users").upsert(row).execute()
            print("[Supabase] Upserted user \(userId) (\(email ?? "no email"))")
        } catch {
            print("[Supabase] Upsert user failed: \(error)")
        }
    }

    func saveFCMToken(userId: String, token: String) async {
        do {
            try await client.from("app_users")
                .update(["fcm_token": token, "platform": "ios"])
                .eq("id", value: userId)
                .execute()
            print("[Supabase] Saved FCM token for user \(userId)")
        } catch {
            print("[Supabase] Save FCM token failed: \(error)")
        }
    }

    // MARK: - Storage

    func uploadVideo(fileURL: URL, generationId: String) async -> String? {
        let storagePath = "renders/\(generationId).mp4"
        do {
            let fileData = try Data(contentsOf: fileURL)
            try await client.storage.from("videos").upload(
                storagePath,
                data: fileData,
                options: .init(contentType: "video/mp4", upsert: true)
            )
            let publicURL = try client.storage.from("videos").getPublicURL(path: storagePath)
            print("[Supabase] Uploaded video: \(publicURL.absoluteString)")
            return publicURL.absoluteString
        } catch {
            print("[Supabase] Upload failed: \(error)")
            return nil
        }
    }

    func deleteVideo(generationId: String) async {
        let storagePath = "renders/\(generationId).mp4"
        do {
            try await client.storage.from("videos").remove(paths: [storagePath])
        } catch {
            print("[Supabase] Delete failed: \(error)")
        }
    }

    // MARK: - Database (generations table)

    struct GenerationRow: Codable {
        let id: String
        let video_name: String
        let video_storage_path: String?
        let status: String
        let created_at: String
        let user_id: String?
        let music_id: String?
        let music_name: String?
        let music_volume: Double?
        let takes_json: String?
        let music_file: String?
    }

    private struct UpsertGenerationRow: Encodable {
        let id: String
        let video_name: String
        let video_storage_path: String?
        let status: String
        let created_at: String
        let user_id: String?
        let music_id: String?
        let music_name: String?
        let music_volume: Double?
        let takes_json: String?
    }

    func upsertGeneration(_ gen: Generation, remoteVideoUrl: String? = nil) async {
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let row = UpsertGenerationRow(
            id: gen.id,
            video_name: gen.videoName,
            video_storage_path: remoteVideoUrl,
            status: gen.status.rawValue,
            created_at: iso.string(from: gen.createdAt),
            user_id: gen.userId,
            music_id: gen.musicId,
            music_name: gen.musicName,
            music_volume: gen.musicVolume,
            takes_json: gen.takesJson
        )
        do {
            try await client.from("generations").upsert(row).execute()
            print("[Supabase] Upserted generation \(gen.id)")
        } catch {
            if isTableNotFoundError(error) {
                print("[Supabase] Generations table not in schema (PGRST205); skipping sync.")
            } else {
                print("[Supabase] Upsert failed: \(error)")
            }
        }
    }

    func fetchGenerations(userId: String) async -> [Generation] {
        do {
            let rows: [GenerationRow] = try await client.from("generations")
                .select()
                .eq("user_id", value: userId)
                .order("created_at", ascending: false)
                .execute()
                .value
            let iso = ISO8601DateFormatter()
            iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            return rows.map { row in
                Generation(
                    id: row.id,
                    videoName: row.video_name,
                    videoUri: nil,
                    resultVideoUrl: row.video_storage_path,
                    status: GenerationStatus(rawValue: row.status) ?? .saved,
                    createdAt: iso.date(from: row.created_at) ?? Date(),
                    userId: row.user_id,
                    musicId: row.music_id,
                    musicName: row.music_name,
                    musicFile: row.music_file,
                    musicVolume: row.music_volume,
                    takesJson: row.takes_json
                )
            }
        } catch {
            if isTableNotFoundError(error) {
                print("[Supabase] Generations table not in schema (PGRST205); using local data only.")
            } else {
                print("[Supabase] Fetch failed: \(error)")
            }
            return []
        }
    }

    func deleteGeneration(id: String) async {
        do {
            try await client.from("generations").delete().eq("id", value: id).execute()
        } catch {
            if !isTableNotFoundError(error) {
                print("[Supabase] Delete row failed: \(error)")
            }
        }
    }

    private func isTableNotFoundError(_ error: Error) -> Bool {
        let msg = "\(error)"
        return msg.contains("Could not find the table") || msg.contains("PGRST205")
    }
}

File: android/app/src/main/java/com/theholylabs/creator/services/SupabaseService.kt
package com.theholylabs.creator.services

import android.util.Log
import com.theholylabs.creator.models.Generation
import com.theholylabs.creator.models.GenerationStatus
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.createSupabaseClient
import io.github.jan.supabase.postgrest.Postgrest
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.storage.Storage
import io.github.jan.supabase.storage.storage
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Supabase client for database and storage.
 * Auth is handled by Firebase — Supabase stores data only.
 */
object SupabaseService {

    private const val TAG = "SupabaseService"

    const val SUPABASE_URL = "https://xxcllmzflnwzqiaslyou.supabase.co"
    const val SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Inh4Y2xsbXpmbG53enFpYXNseW91Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzQyMDM1NDcsImV4cCI6MjA4OTc3OTU0N30.VbAv5UEC6n1ZyXFvjLshVOp9ZKN5_Rq8fEznoEqFsKY"

    val client: SupabaseClient by lazy {
        createSupabaseClient(SUPABASE_URL, SUPABASE_ANON_KEY) {
            install(Postgrest)
            install(Storage)
        }
    }

    // ── App Users (Firebase UID + FCM token) ─────────────────────────────

    // ── App Users ─────────────────────────────────────────────────────────

    @Serializable
    private data class AppUserRow(
        val id: String,
        val email: String? = null,
        @SerialName("display_name") val displayName: String? = null,
        @SerialName("avatar_url") val avatarUrl: String? = null,
        @SerialName("fcm_token") val fcmToken: String? = null,
        val platform: String = "android"
    )

    suspend fun upsertUser(userId: String, email: String?, displayName: String? = null, avatarUrl: String? = null) {
        val row = AppUserRow(id = userId, email = email, displayName = displayName, avatarUrl = avatarUrl)
        try {
            client.from("app_users").upsert(row)
            Log.d(TAG, "Upserted user $userId ($email)")
        } catch (e: Exception) {
            Log.e(TAG, "Upsert user failed: ${e.message}")
        }
    }

    suspend fun saveFCMToken(userId: String, token: String) {
        try {
            client.from("app_users").update(
                mapOf("fcm_token" to token, "platform" to "android")
            ) {
                filter { eq("id", userId) }
            }
            Log.d(TAG, "Saved FCM token for user $userId")
        } catch (e: Exception) {
            Log.e(TAG, "Save FCM token failed: ${e.message}")
        }
    }

    // ── Storage ──────────────────────────────────────────────────────────

    private const val VIDEO_BUCKET = "videos"

    suspend fun uploadVideo(videoBytes: ByteArray, generationId: String): String? {
        val storagePath = "renders/$generationId.mp4"
        return try {
            client.storage.from(VIDEO_BUCKET).upload(storagePath, videoBytes, upsert = true)
            val publicUrl = client.storage.from(VIDEO_BUCKET).publicUrl(storagePath)
            Log.d(TAG, "Uploaded video: $publicUrl")
            publicUrl
        } catch (e: Exception) {
            Log.e(TAG, "Upload failed: ${e.message}")
            null
        }
    }

    suspend fun deleteVideo(generationId: String) {
        val storagePath = "renders/$generationId.mp4"
        try {
            client.storage.from(VIDEO_BUCKET).delete(storagePath)
        } catch (e: Exception) {
            Log.e(TAG, "Delete video failed: ${e.message}")
        }
    }

    // ── Database (generations table) ────────────────────────────────────

    @Serializable
    data class GenerationRow(
        val id: String,
        val video_name: String,
        val video_storage_path: String? = null,
        val status: String,
        val created_at: String,
        val user_id: String? = null,
        val music_id: String? = null,
        val music_name: String? = null,
        val music_volume: Double? = null,
        val takes_json: String? = null,
        val music_file: String? = null
    )

    @Serializable
    private data class UpsertRow(
        val id: String,
        val video_name: String,
        val video_storage_path: String? = null,
        val status: String,
        val created_at: String,
        val user_id: String? = null
    )

    suspend fun upsertGeneration(gen: Generation, remoteVideoUrl: String? = null) {
        val row = UpsertRow(
            id = gen.id,
            video_name = gen.videoName,
            video_storage_path = remoteVideoUrl ?: gen.resultVideoUrl,
            status = gen.status.name.lowercase(),
            created_at = gen.createdAt,
            user_id = gen.userId
        )
        try {
            client.from("generations").upsert(row)
            Log.d(TAG, "Upserted generation ${gen.id}")
        } catch (e: Exception) {
            if (isTableNotFound(e)) {
                Log.w(TAG, "Generations table not in schema; skipping sync.")
            } else {
                Log.e(TAG, "Upsert failed: ${e.message}")
            }
        }
    }

    suspend fun fetchGenerations(userId: String): List<Generation> {
        return try {
            val rows: List<GenerationRow> = client.from("generations")
                .select {
                    filter { eq("user_id", userId) }
                    order("created_at", io.github.jan.supabase.postgrest.query.Order.DESCENDING)
                }
                .decodeList()

            rows.map { row ->
                Generation(
                    id = row.id,
                    videoName = row.video_name,
                    videoUri = null,
                    resultVideoUrl = row.video_storage_path,
                    status = try {
                        GenerationStatus.valueOf(row.status.uppercase())
                    } catch (_: Exception) {
                        GenerationStatus.SAVED
                    },
                    createdAt = row.created_at,
                    userId = row.user_id
                )
            }
        } catch (e: Exception) {
            if (isTableNotFound(e)) {
                Log.w(TAG, "Generations table not in schema; using local data only.")
            } else {
                Log.e(TAG, "Fetch failed: ${e.message}")
            }
            emptyList()
        }
    }

    suspend fun deleteGeneration(id: String) {
        try {
            client.from("generations").delete {
                filter { eq("id", id) }
            }
        } catch (e: Exception) {
            if (!isTableNotFound(e)) {
                Log.e(TAG, "Delete row failed: ${e.message}")
            }
        }
    }

    private fun isTableNotFound(error: Exception): Boolean {
        val msg = error.message ?: ""
        return msg.contains("Could not find the table") || msg.contains("PGRST205")
    }
}
